### PR TITLE
[PR 1] DEV-177 BLCU fixes

### PR DESCRIPTION
**Pull request by jmaralo**
After integrating the BLCU logic with FW, here are the fixes necessary to make it work, it also features an update to notify the progress to the frontend. However, This is currently not supported by the frontend and will cause issues.

### CODE REVIEW

File: src/blcu/download.go [hunk 1]
```diff
@@ -75,43 +75,50 @@ func (blcu *BLCU) createDownloadOrder(board string) models.Order {
...
 	if err != nil {
-		return err
+		return Download{}, err
 	}
 
-	_, err = receiver.WriteTo(writer)
-	if err != nil {
-		return err
-	}
+	download := NewDownload(786432) // Size of the flash memory (downlaod size)
```
Comment (smorsua) [inline src/blcu/download.go:91]: Put size in const FlashMemorySize

File: src/blcu/download.go [hunk 2]
```diff
@@ -17,34 +17,34 @@ type downloadRequest struct {
...
 		blcu.trace.Error().Err(err).Stack().Msg("Request download")
-		return downloadData{}, err
+		return request.Board, DownloadResult{Err: err}
 	}
 
-	buffer := bytes.NewBuffer([]byte{})
-	if err := blcu.ReadTFTP(buffer); err != nil {
-		blcu.trace.Error().Err(err).Stack().Msg("Read TFTP")
-		return downloadData{}, err
+	download, err := blcu.ReadTFTP()
```
Comment (smorsua) [inline src/blcu/download.go:33]: I would do the following: ReadTFTP blocks until it downloads or errors and it accepts a callback (onProgress), where you receive the progress and can whatever you want to do with it.

File: src/blcu/download.go [hunk 3]
```diff
@@ -75,43 +75,50 @@ func (blcu *BLCU) createDownloadOrder(board string) models.Order {
...
 
-	return nil
+	return download, nil
 }
 
 type downloadResponse struct {
-	IsSuccess bool   `json:"success"`
-	File      []byte `json:"file,omitempty"`
+	Percentage float64 `json:"percentage"`
+	IsSuccess  *bool   `json:"success,omitempty"`
```
Comment (smorsua) [inline src/blcu/download.go:99]: Make IsSuccess value and send false while percentage is less than 100

File: src/blcu/download.go [hunk 4]
```diff
@@ -123,5 +130,80 @@ func (blcu *BLCU) writeDownloadFile(data downloadData) error {
...
-	return os.WriteFile(path.Join(blcu.config.DownloadPath, data.Board+".bin"), data.Payload, 0777)
+	return os.WriteFile(path.Join(blcu.config.DownloadPath, fmt.Sprintf("%s-%d.bin", board, time.Now().Unix())), data, 0777)
+}
+
+type DownloadResult struct {
+	Data []byte
+	Err  error
+}
+
+type Download struct {
```
Comment (smorsua) [inline src/blcu/download.go:141]: Download would store the onProgress callback passed to ReadTFTP

File: src/blcu/download.go [hunk 5]
```diff
@@ -123,5 +130,80 @@ func (blcu *BLCU) writeDownloadFile(data downloadData) error {
...
+func (download *Download) Write(p []byte) (n int, err error) {
+	if download.writeErr != nil {
+		return 0, download.writeErr
+	}
+	n, err = download.output.Write(p)
+	if err != nil {
+		download.writeErr = err
+		download.abort(err)
+	} else {
+		download.notify <- n
```
Comment (smorsua) [inline src/blcu/download.go:181]: You'd call onProgress here, with the percentage

File: src/blcu/download.go [hunk 6]
```diff
@@ -123,5 +130,80 @@ func (blcu *BLCU) writeDownloadFile(data downloadData) error {
...
+	}
+	return n, err
+}
+
+func (download *Download) abort(err error) {
```
Comment (smorsua) [inline src/blcu/download.go:186]: Why do you need to send the error through a channel? receiver.WriteTo returns an error, so consumers should check that error to determine if the download failed

File: src/blcu/download.go [hunk 7]
```diff
@@ -123,5 +130,80 @@ func (blcu *BLCU) writeDownloadFile(data downloadData) error {
...
+	close(download.notify)
+	download.resultChan <- DownloadResult{
```
Comment (smorsua) [inline src/blcu/download.go:188]: If receiver.WriteTo block, you dont need this channel

**Review by smorsua — CHANGES_REQUESTED** [review]
Pretty clear implementation, I'd avoid concurrency in some places to avoid using many channels and such

**Review by smorsua — APPROVED** [review]
LGTM

**Review by lipezaballa — APPROVED** [review]

**Pull request merged**